File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
#from collections import deque
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        # ans=[]
        # if not root:
        #     return root
        # def dfs(head):
        #     if head:
        #         ans.append(head.val)
        #         dfs(head.left)
        #         dfs(head.right)
        # head=root        
        # dfs(head)
        # root=TreeNode(ans[0])
        # for i in range(1,len(ans)):
        #     root.right=TreeNode(ans[i])
        #     #print(root.right.val)
        #     root=root.right
        if not root:
            return root
        stack=[]
        stack.append(root)
        while stack:
            aux=stack.pop()
            if aux.right:
                stack.append(aux.right)
            if aux.left:
                stack.append(aux.left)
            if stack:    
                aux.right=stack[-1]
                aux.left=None
            
        
            
        
         
           
            
        
        

File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // use a previous pointer to maintain the prev node
    TreeNode* prev=NULL;
    void flatten(TreeNode* root) {
        if(root!=NULL){
            // first visit the right child
            //then visit the left child
            flatten(root->right);
            flatten(root->left);
            //make left point to NULL
            
            root->left=NULL;
            //right to prev
            root->right=prev;
            //change prev to root
            prev=root;
        }
        
    }
};
